### PR TITLE
fix(reporter): Filter out scan results outside the VCS path

### PR DESCRIPTION
**Pull request by wkl3nk**
The scanner does not take care for the VCS path of a repository provenance, it just scans **everything** and puts the scan results to persistent storage (scanner cache). This is the desired behaviour and nothing wrong about it.

But when scan results from this persistent storage are used, scanner findings outside the VCS path must be filtered before further processing.

For solving the issue, the first step was to isolate the respective code sections in ORT to a Util class.
With this change, re-use this code from the Util class in ORT Server in order to filter out scan results outside the VCS path. This ensures that ORT Server behaves in exactly the same way as ORT regarding scan results.

### CODE REVIEW

**Review by oheger-bosch — COMMENTED** [review]
The changes look good to me, but I am missing a test.

File: workers/common/src/main/kotlin/common/OrtRunService.kt [hunk 1]
```diff
@@ -489,4 +499,32 @@ class OrtRunService(
...
+     * For scan results regarding a [RepositoryProvenance], if this provenance has a VCS path, then filter
+     * out the [OrtScanResult] summaries that are not related to this path. These summaries typically contain
+     * findings about licenses, copyrights, code snippets and other information from the scan.
+     */
+    private fun filterScanResultSummary(
+        provenances: Collection<ProvenanceResolutionResult>,
+        scanResults: Set<ScanResult>
+    ): Set<OrtScanResult> {
+        val vcsPaths = provenances
+            ?.mapNotNull { (it.packageProvenance as? RepositoryProvenance)?.vcsInfo?.path }
```
Comment (sschuberth) [inline workers/common/src/main/kotlin/common/OrtRunService.kt:513]: `provenances` is not nullable, so the Elvis-operators are not needed here, allowing to simplify code. For readability maybe also consider filtering for `RepositoryProvenance` instances separately, but together with the mapping that would be two times iterating `provenances`, so maybe not desirable.

File: workers/common/src/main/kotlin/common/OrtRunService.kt [hunk 2]
```diff
@@ -489,4 +499,32 @@ class OrtRunService(
...
+            ?.filter { it.isNotEmpty() }
+            ?.toMutableSet() ?: mutableSetOf()
```
Comment (sschuberth) [inline workers/common/src/main/kotlin/common/OrtRunService.kt:515]: Actually, there's no need for `vcsPaths` to be mutable.

File: workers/common/src/main/kotlin/common/OrtRunService.kt [hunk 3]
```diff
@@ -489,4 +499,32 @@ class OrtRunService(
...
+
+        // Filter out scan result summaries outside the VCS path.
+        return scanResults
+            ?.map { scanResult ->
```
Comment (sschuberth) [inline workers/common/src/main/kotlin/common/OrtRunService.kt:519]: Also here, no need for the Elvis-operator.

File: workers/common/src/main/kotlin/common/OrtRunService.kt [hunk 4]
```diff
@@ -489,4 +499,32 @@ class OrtRunService(
...
+                scanResult.mapToOrt().let { ortScanResult ->
+                    if (scanResult.provenance is RepositoryProvenance) {
+                        ortScanResult.copy(summary = ortScanResult.summary.filterByPaths(vcsPaths))
+                    } else {
+                        ortScanResult
+                    }
+                }
+            }
+            ?.toSet().orEmpty()
```
Comment (sschuberth) [inline workers/common/src/main/kotlin/common/OrtRunService.kt:528]: Directly `mapTo` a set instead.

File: workers/common/src/main/kotlin/common/OrtRunService.kt [hunk 5]
```diff
@@ -489,4 +499,32 @@ class OrtRunService(
...
+    }
```
Comment (sschuberth) [inline workers/common/src/main/kotlin/common/OrtRunService.kt:529]: All in all, I propose instead:

        val vcsPaths = provenances
            .filterIsInstance<RepositoryProvenance>()
            .mapNotNull { it.vcsInfo.path.takeUnless { it.isEmpty() } }

        // Filter out scan result summaries outside the VCS path.
        return scanResults.mapTo(mutableSetOf()) { scanResult ->
            scanResult.mapToOrt().let { ortScanResult ->
                if (scanResult.provenance is RepositoryProvenance) {
                    ortScanResult.copy(summary = ortScanResult.summary.filterByPaths(vcsPaths))
                } else {
                    ortScanResult
                }
            }
        }

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by sschuberth — DISMISSED** [review]

File: workers/common/src/main/kotlin/common/OrtRunService.kt [hunk 6]
```diff
@@ -489,4 +499,29 @@ class OrtRunService(
...
+        // Filter out scan result summaries outside the VCS path.
+        return scanResults.mapTo(mutableSetOf()) { scanResult ->
+            scanResult.mapToOrt().let { ortScanResult ->
+                if (scanResult.provenance is RepositoryProvenance) {
+                    ortScanResult.copy(summary = ortScanResult.summary.filterByPaths(vcsPaths))
+                } else {
+                    ortScanResult
+                }
+            }
+        }
```
Comment (mnonnenmacher) [inline workers/common/src/main/kotlin/common/OrtRunService.kt:525]: If I understand this function correctly it collects the VCS paths from all repository provenances in the scanner run in `vcsPaths` and then filters all scan results with a repository provenance by these paths.

So if repo A has VCS path "a", all scan results get filtered by path "a" no matter if they from repo A or any other repo?

Compare this to this code from ORT which does collects the VCS paths by provenance and then filter each scan result only by the paths that match the provenance:
https://github.com/oss-review-toolkit/ort/blob/254dbf96933d9514add566c62df5e41564a5612b/scanner/src/main/kotlin/Scanner.kt#L194-L212

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: workers/common/src/main/kotlin/common/OrtRunService.kt (hunk 6 above)
Comment (wkl3nk) [reply workers/common/src/main/kotlin/common/OrtRunService.kt:525]: All in all the situation is not nice. If ORT Server sees itself as a wrapper around ORT, then ORT Server should not implement its own logic to work on the scan results, but instead use the same logic as in ORT. Okay, I can have at the code in ORT, adapt it and make a copy of the code in ORT Server that does exactly the same, but ... not nice.

Comment (mnonnenmacher) [reply workers/common/src/main/kotlin/common/OrtRunService.kt:525]: It would be better to move the code in ORT into a reusable function than to copy it.

Comment (wkl3nk) [reply workers/common/src/main/kotlin/common/OrtRunService.kt]: Did it this way now. Isolated the filter code in ORT into a Util class, and now just re-using the same in ORT Server to filter out the unwanted scan results, in the same way it is already done in ORT.

**Comment by wkl3nk** [comment]
ort-result.zip section before change for a test project that used only VCS path _npm/simple_ . You can see paths different from _npm/simple_ in the scan results:

```
  scan_results:
  - provenance:
      vcs_info:
        type: "Git"
        url: "https://github.com/bosch-ocaas/ocaas-test-projects.git"
        revision: "05f320aff2e2b565f150a48a4f801382f6cf5987"
        path: ""
      resolved_revision: "05f320aff2e2b565f150a48a4f801382f6cf5987"
    scanner:
      name: "ScanCode"
      version: "32.2.1"
      configuration: "--copyright --license --info --strip-root --timeout 300 --max-in-memory\
        \ 5000 --json-pp"
    summary:
      start_time: "2024-10-16T15:10:56Z"
      end_time: "2024-10-16T15:11:08Z"
      licenses:
      - license: "Apache-2.0"
        location:
          path: "gradle-inspector/bootstrap-java17/.ort.yml"
          start_line: 20
          end_line: 20
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle-inspector/bootstrap-java17/gradlew"
          start_line: 6
          end_line: 16
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle-inspector/bootstrap-java17/gradlew"
          start_line: 18
          end_line: 18
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle-inspector/bootstrap-java17/gradlew.bat"
          start_line: 4
          end_line: 14
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle-inspector/bootstrap-java17/gradlew.bat"
          start_line: 16
          end_line: 16
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle-inspector/bootstrap-java21/.ort.yml"
          start_line: 20
          end_line: 20
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle-inspector/bootstrap-java21/gradlew"
          start_line: 6
          end_line: 16
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle-inspector/bootstrap-java21/gradlew"
          start_line: 18
          end_line: 18
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle-inspector/bootstrap-java21/gradlew.bat"
          start_line: 4
          end_line: 14
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle-inspector/bootstrap-java21/gradlew.bat"
          start_line: 16
          end_line: 16
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle/bootstrap-java17/.ort.yml"
          start_line: 20
          end_line: 20
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle/bootstrap-java17/gradlew"
          start_line: 6
          end_line: 16
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle/bootstrap-java17/gradlew"
          start_line: 18
          end_line: 18
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle/bootstrap-java17/gradlew.bat"
          start_line: 4
          end_line: 14
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle/bootstrap-java17/gradlew.bat"
          start_line: 16
          end_line: 16
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle/bootstrap-java21/.ort.yml"
          start_line: 20
          end_line: 20
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle/bootstrap-java21/gradlew"
          start_line: 6
          end_line: 16
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle/bootstrap-java21/gradlew"
          start_line: 18
          end_line: 18
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle/bootstrap-java21/gradlew.bat"
          start_line: 4
          end_line: 14
        score: 100.0
      - license: "Apache-2.0"
        location:
          path: "gradle/bootstrap-java21/gradlew.bat"
          start_line: 16
          end_line: 16
        score: 100.0
      - license: "BSD-3-Clause"
        location:
          path: "npm/simple/package.json"
          start_line: 4
          end_line: 4
        score: 100.0
      - license: "GPL-3.0-only"
        location:
          path: "npm/simple/package.json"
          start_line: 4
          end_line: 4
        score: 100.0
      - license: "ISC"
        location:
          path: "npm/simple/package.json"
          start_line: 6
          end_line: 6
        score: 100.0
      - license: "MIT"
        location:
          path: "npm/simple/package.json"
          start_line: 4
          end_line: 4
        score: 100.0
      copyrights:
      - statement: "Copyright (c) 2015-2021 the original"
        location:
          path: "gradle-inspector/bootstrap-java17/gradlew"
          start_line: 4
          end_line: 4
      - statement: "Copyright (c) 2015-2021 the original"
        location:
          path: "gradle-inspector/bootstrap-java21/gradlew"
          start_line: 4
          end_line: 4
      - statement: "Copyright (c) 2015-2021 the original"
        location:
          path: "gradle/bootstrap-java17/gradlew"
          start_line: 4
          end_line: 4
      - statement: "Copyright (c) 2015-2021 the original"
        location:
          path: "gradle/bootstrap-java21/gradlew"
          start_line: 4
          end_line: 4
      - statement: "Copyright 2015 the original author or authors"
        location:
          path: "gradle-inspector/bootstrap-java17/gradlew.bat"
          start_line: 2
          end_line: 2
      - statement: "Copyright 2015 the original author or authors"
        location:
          path: "gradle-inspector/bootstrap-java21/gradlew.bat"
          start_line: 2
          end_line: 2
      - statement: "Copyright 2015 the original author or authors"
        location:
          path: "gradle/bootstrap-java17/gradlew.bat"
          start_line: 2
          end_line: 2
      - statement: "Copyright 2015 the original author or authors"
        location:
          path: "gradle/bootstrap-java21/gradlew.bat"
          start_line: 2
          end_line: 2
```

ort-result.zip section after change for a test project that used only VCS path _npm/simple_ . 


```
  - provenance:
      vcs_info:
        type: "Git"
        url: "https://github.com/bosch-ocaas/ocaas-test-projects.git"
        revision: "05f320aff2e2b565f150a48a4f801382f6cf5987"
        path: ""
      resolved_revision: "05f320aff2e2b565f150a48a4f801382f6cf5987"
    scanner:
      name: "ScanCode"
      version: "32.2.1"
      configuration: "--copyright --license --info --strip-root --timeout 300 --max-in-memory\
        \ 5000 --json-pp"
    summary:
      start_time: "2024-10-16T15:10:56Z"
      end_time: "2024-10-16T15:11:08Z"
      licenses:
      - license: "BSD-3-Clause"
        location:
          path: "npm/simple/package.json"
          start_line: 4
          end_line: 4
        score: 100.0
      - license: "GPL-3.0-only"
        location:
          path: "npm/simple/package.json"
          start_line: 4
          end_line: 4
        score: 100.0
      - license: "ISC"
        location:
          path: "npm/simple/package.json"
          start_line: 6
          end_line: 6
        score: 100.0
      - license: "MIT"
        location:
          path: "npm/simple/package.json"
          start_line: 4
          end_line: 4
        score: 100.0
```

**Comment by oheger-bosch** [comment]
Still missing a test.

**Comment by wkl3nk** [comment]
@wkl3nk The filtering now is isolated in a function to make it testable. Added 2 simple tests to make sure the scan results remain untouched when no _VCSPath_ is used at all, but that it works if actually a _VCSPath_ is used.

**Review by oheger-bosch — APPROVED** [review]

**Review by mnonnenmacher — APPROVED** [review]

**Pull request merged**